includes/Transport/HttpTransport.php

```diff
@@ -0,0 +1,497 @@
+	 * @param mixed $request The request object.
+	 *
+	 * @return array
+	 */
+	private function handle_initialize_request( array $message, $request ): array {
+		$request_id = (int) $message['id'];
+		$params     = $message['params'] ?? array();
+
+		// Route the initialize request
+		$result = $this->context->request_router->route_request(
+			'initialize',
+			$params,
+			$request_id,
+			$this->get_transport_name()
+		);
+
+		// If successful, create a session for all clients (unified session management)
+		if ( ! isset( $result['error'] ) ) {
+			// Check if client already has a session
+			$existing_session = $request->get_header( 'Mcp-Session-Id' );
+
+			if ( ! $existing_session ) {
+				// Create new session for all clients (proxy and direct)
+				$session_id = $this->create_session( $params['clientInfo'] ?? array() );
+
+				// Add session header to response
+				add_filter(
+					'rest_post_dispatch',
+					static function ( $response ) use ( $session_id ) {
+						if ( $response instanceof WP_REST_Response ) {
+							$response->header( 'Mcp-Session-Id', $session_id );
+						}
+						return $response;
+					}
+				);
+			}
+		}
+
+		// Format as JSON-RPC response
+		if ( isset( $result['error'] ) ) {
+			return array(
+				'jsonrpc' => '2.0',
+				'id'      => $request_id,
+				'error'   => $result['error'],
+			);
+		}
+
+		return array(
+			'jsonrpc' => '2.0',
+			'id'      => $request_id,
+			'result'  => $result,
+		);
+	}
+
+
+
+	/**
+	 * Validate session for all client requests (unified session management)
+	 *
+	 * @param mixed $request The request object.
+	 *
+	 * @return array|true Returns true if valid, error array if invalid.
+	 */
+	private function validate_session( $request ) {
+		$session_id = $request->get_header( 'Mcp-Session-Id' );
+
+		if ( ! $session_id ) {
+			return McpErrorFactory::invalid_request( 0, 'Missing Mcp-Session-Id header' );
+		}
+
+		// All sessions are now managed by WordPress - check transients
+		$session_data = get_transient( "mcp_session_{$session_id}" );
+
+		if ( false === $session_data ) {
+			return McpErrorFactory::invalid_request( 0, 'Invalid or expired session' );
+		}
+
+		// Session found - update last activity and refresh expiration
+		$session_data['last_activity'] = time();
+		set_transient( "mcp_session_{$session_id}", $session_data, self::SESSION_EXPIRATION );
+
+		return true;
+	}
+
+	/**
+	 * Create a new session using WordPress transients
+	 *
+	 * @param array $client_info The client information from initialize request.
+	 *
+	 * @return string The session ID.
+	 */
+	private function create_session( array $client_info ): string {
+		$session_id = wp_generate_uuid4();
+
+		$session_data = array(
+			'created_at'    => time(),
+			'last_activity' => time(),
+			'client_info'   => $client_info,
+		);
+
+		// Store session as transient with automatic expiration
+		set_transient( "mcp_session_{$session_id}", $session_data, self::SESSION_EXPIRATION );
+
+		return $session_id;
+	}
```

Comment:
I'm not sure putting MCP sessions into transients like this is the best approach. Here, I'm vary of DoS attacks. Every time a new request is made, a new transient is added, which can quickly flood the database or object cache.

It would be interesting to find a way to tie a session ID to a specific user. For example through user meta or so. Could be a good for additional hardening and allows for a quick way to remove all sessions by a certain user (like regular user sessions)

In https://github.com/mcp-wp/mcp-server I created a custom post type, but that's not ideal either.